library/ssl_tls13_client.c

```diff
@@ -3303,6 +3283,26 @@ static int ssl_server_hello_postprocess( mbedtls_ssl_context* ssl )
     }
 #endif /* MBEDTLS_CID */
 
+    mbedtls_ssl_key_set traffic_keys;
+
+    /* Generate handshake keying material */
+    ret = mbedtls_ssl_handshake_key_derivation( ssl, &traffic_keys );
+    if( ret != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_handshake_key_derivation", ret );
+        return( ret );
+    }
+
+    ret = mbedtls_ssl_tls13_build_transform( ssl, &traffic_keys, ssl->transform_handshake, 0 );
+    if( ret != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_tls13_build_transform", ret );
+        return( ret );
+    }
+
+    /* Switch to new keys for inbound traffic. */
+    mbedtls_ssl_set_inbound_transform( ssl, ssl->transform_handshake );
+    ssl->session_in = ssl->session_negotiate;
```

Comment:
Pre-existing defect worth fixing along the way: We must zeroize the sensitive stack variable `mbedtls_ssl_traffic_keys` here, using `mbedtls_platform_zeroize()`.